### PR TITLE
feat: byob: secure upload attestation and public download Action

### PR DESCRIPTION
**Pull request by laurentsimon**
closes https://github.com/slsa-framework/slsa-github-generator/issues/1571

### CODE REVIEW

File: actions/delegator/secure-attestations-download/action.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+name: "Secure attestations download"
+description: "Download a set of attestations and verify their SHA256"
```
Comment (laurentsimon) [inline actions/delegator/secure-attestations-download/action.yml:2]: is this path acceptable for a public Action? Will "delegator" be confusing to end-users?

File: actions/nodejs/secure-attestations-download/action.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+name: "Secure attestations download"
```
Comment (laurentsimon) [inline actions/nodejs/secure-attestations-download/action.yml:1]: Creating a dedicated Action that matches the "nodejs" builder end-users will be familiar with. I think this also allows us to update a builder without breaking others.

File: actions/nodejs/secure-attestations-download/action.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+name: "Secure attestations download for nodejs builder"
```
Comment (ianlewis) [inline actions/nodejs/secure-attestations-download/action.yml:1]: I realize we only have the nodejs builder right now but I think most BYOB workflows could use this, no?

Also what situation were we expecting folks to use this? I guess if they need to do npm publish on their own?

File: .github/workflows/scripts/pre-submit.delegators/expected.diff [hunk 4]
```diff
@@ -0,0 +1,16 @@
+15c15,19
+< name: SLSA builder delegator
```
Comment (laurentsimon) [inline .github/workflows/scripts/pre-submit.delegators/expected.diff:2]: ignore this file, I need to rebase

File: actions/nodejs/secure-attestations-download/action.yml (hunk 3 above)
Comment (laurentsimon) [reply actions/nodejs/secure-attestations-download/action.yml:1]: yes all folks can use it. I'm hesitant to create one for all ecosystems. If we ever need to change this, at least we'll be able to break only one of the builders, not all.

Comment (ianlewis) [reply actions/nodejs/secure-attestations-download/action.yml:1]: Yeah, it's fine for now I guess.

File: .github/workflows/builder_nodejs_slsa3.yml [hunk 5]
```diff
@@ -108,7 +108,7 @@ jobs:
       id-token: write # For signing.
       contents: read # For repo checkout of private repos.
       actions: read # For getting workflow run on private repos.
-    uses: slsa-framework/slsa-github-generator/.github/workflows/delegator_lowperms-generic_slsa3.yml@main
+    uses: laurentsimon/slsa-github-generator/.github/workflows/delegator_lowperms-generic_slsa3.yml@feat/sec-upload-att
```
Comment (ianlewis) [inline .github/workflows/builder_nodejs_slsa3.yml:111]: FIXME

File: .github/workflows/builder_nodejs_slsa3.yml [hunk 6]
```diff
@@ -127,9 +127,10 @@ jobs:
       # NOTE: The attestation name has a rng value in it's name so we don't need to check for file collisions.
       - name: Download attestations
         id: attestation-download
-        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
+        uses: laurentsimon/slsa-github-generator/actions/nodejs/secure-attestations-download@feat/sec-upload-att
```
Comment (ianlewis) [inline .github/workflows/builder_nodejs_slsa3.yml:130]: FIXME

File: .github/workflows/builder_nodejs_slsa3.yml [hunk 7]
```diff
@@ -127,9 +127,11 @@ jobs:
       # NOTE: The attestation name has a rng value in it's name so we don't need to check for file collisions.
       - name: Download attestations
         id: attestation-download
-        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
+        uses: slsa-framework/slsa-github-generator/actions/nodejs/secure-attestations-download@main
         with:
```
Comment (laurentsimon) [inline .github/workflows/builder_nodejs_slsa3.yml:131]: note: end-uses will need to use a similar command. We need to tell them that the name of the folder containing all the attestations is `${{ needs.slsa-run.outputs.attestations-download-name }}`.

@ianlewis is that OK in terms of UX?

Comment (laurentsimon) [reply .github/workflows/builder_nodejs_slsa3.yml:131]: you said you'll provide an Action as well to simplify

Comment (ianlewis) [reply .github/workflows/builder_nodejs_slsa3.yml:131]: Yeah, this is fine. I have some plans for simplification.

I'd like to make the artifact uploads and the file names in the upload match if possible, and provide an action for publishing to make it easier.

**Review by ianlewis — APPROVED** [review]
No delegator diff change?

**Comment by laurentsimon** [comment]
> No delegator diff change?

apparently not.. because I changed the exact same lines in both workflows so they match.

**Comment by laurentsimon** [comment]
> > No delegator diff change?
> 
> apparently not.. because I changed the exact same lines in both workflows so they match.

And I'm obviously wrong. pre-submit is failing :-)

**Comment by laurentsimon** [comment]
I marked the delegator pre-submit as required in branch protection settings.

**Pull request merged**